packages/react-i18n/src/index.tsx

```diff
@@ -1,3 +1,10 @@
+/**
+ * External dependencies
+ */
+// Disable reason: Type-only import, this is fine. See https://github.com/typescript-eslint/typescript-eslint/issues/2661
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType, PropsWithChildren } from 'react';
```

Comment:
I agree up to a point.

We looked at doing that and it was a nightmare at the time and would've required a lot of maintenance effort for little or no benefit. There's no way to work nicely with types in the module import/export system in JSDoc. We would've had to manually re-export every single type (with their type arguments!) one by one via JSDoc.

Now that we're allowing some TypeScript, it may be more feasible to re-export all the types, but I haven't looked in many months. It still won't be as easy as something like `export type * from 'react'` as far as I know.

Reply:
Unless there is some way to let ESLint know that it should give itself a break when seeing `import type` when processing `no-restricted-imports` 🤔 

Let's see: https://github.com/eslint/eslint/issues/13758.

Reply:
Maybe we should list React as peer dependency in that case? There are more packages that would require the same handling.

Reply:
I reviewed all the changes again.

It won't scale with so many types required from React or Reakit 🙁

Should we at least disable this rule for all `packages/**/types.ts` files? Ideally, the rule shouldn't error for type imports in the first place.

Reply:
I think we should just disable those rules for `packages/components`... G2 imports a _lot_ from reakit as well, outside of typescript files.

Reply:
We could as well remove those checks for `react` and `reakit` if that becomes an issue. Although, my initial comment still applies. Technically speaking both those libraries are implementation details and shouldn't leak outside of `@wordpress/element` (`react`) and `@wordpress/components` (`reakit`). If there is no way to re-export the types that are used in other packages, then we shouldn't fight with the linter and disable the rule as a primary solution. We better remove the rule that becomes hostile.